File: apps/android/app/src/main/java/com/nous/app/data/NousApi.kt
package com.nous.app.data

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class HealthResponse(
    val status: String,
    val version: String,
    val uptime_ms: Long,
)

@Serializable
data class IdentityResponse(
    val did: String,
    val display_name: String? = null,
    val signing_key_type: String,
    val exchange_key_type: String,
)

@Serializable
data class BalanceEntry(
    val token: String,
    val amount: String,
)

@Serializable
data class WalletResponse(
    val did: String,
    val balances: List<BalanceEntry>,
    val nonce: Int,
    val created_at: String,
)

@Serializable
data class FeedEvent(
    val id: String,
    val pubkey: String,
    val created_at: String,
    val kind: Int,
    val content: String,
    val tags: List<List<String>>,
)

@Serializable
data class FeedResponse(
    val events: List<FeedEvent>,
    val count: Int,
)

@Serializable
data class DaoResponse(
    val id: String,
    val name: String,
    val description: String,
    val founder_did: String,
    val member_count: Int,
    val created_at: String,
)

@Serializable
data class DaoListResponse(
    val daos: List<DaoResponse>,
    val count: Int,
)

@Serializable
data class ProposalResponse(
    val id: String,
    val dao_id: String,
    val title: String,
    val description: String,
    val proposer_did: String,
    val status: String,
    val created_at: String,
    val voting_starts: String,
    val voting_ends: String,
    val quorum: Double,
    val threshold: Double,
)

@Serializable
data class ProposalListResponse(
    val proposals: List<ProposalResponse>,
    val count: Int,
)

@Serializable
data class TransactionResponse(
    val id: String,
    val from_did: String,
    val to_did: String,
    val token: String,
    val amount: String,
    val memo: String? = null,
    val status: String,
    val created_at: String,
)

@Serializable
data class TransactionListResponse(
    val transactions: List<TransactionResponse>,
    val count: Int,
)

@Serializable
data class ChannelResponse(
    val id: String,
    val name: String,
    val channel_type: String,
    val member_count: Int,
    val last_message: String? = null,
    val last_message_at: String? = null,
    val created_at: String,
)

@Serializable
data class ChannelListResponse(
    val channels: List<ChannelResponse>,
    val count: Int,
)

@Serializable
data class MessageResponse(
    val id: String,
    val channel_id: String,
    val sender_did: String,
    val content: String,
    val created_at: String,
)

@Serializable
data class MessageListResponse(
    val messages: List<MessageResponse>,
    val count: Int,
)

@Serializable
data class SendMessageRequest(
    val content: String,
)

@Serializable
data class SendTransactionRequest(
    val to_did: String,
    val token: String,
    val amount: String,
    val memo: String? = null,
)

@Serializable
data class CreatePostRequest(
    val content: String,
    val tags: List<String> = emptyList(),
)

@Serializable
data class CreateIdentityRequest(
    val display_name: String? = null,
)

class NousApi(private val baseUrl: String = "http://10.0.2.2:8080/api/v1") {

    private val client = HttpClient(OkHttp) {
        install(ContentNegotiation) {
            json(Json {
                ignoreUnknownKeys = true
                isLenient = true
            })
        }
    }

    suspend fun health(): HealthResponse =
        client.get("$baseUrl/health").body()

    suspend fun createIdentity(displayName: String? = null): IdentityResponse =
        client.post("$baseUrl/identities") {
            contentType(ContentType.Application.Json)
            setBody(CreateIdentityRequest(displayName))
        }.body()

    suspend fun getIdentity(did: String): IdentityResponse =
        client.get("$baseUrl/identities/$did").body()

    suspend fun getWallet(did: String): WalletResponse =
        client.get("$baseUrl/wallets/$did").body()

    suspend fun getFeed(limit: Int = 20): FeedResponse =
        client.get("$baseUrl/feed?limit=$limit").body()

    suspend fun listDaos(): DaoListResponse =
        client.get("$baseUrl/daos").body()

    suspend fun listProposals(): ProposalListResponse =
        client.get("$baseUrl/proposals").body()

    suspend fun getTransactions(did: String, limit: Int = 50): TransactionListResponse =
        client.get("$baseUrl/wallets/$did/transactions?limit=$limit").body()

    suspend fun sendTransaction(fromDid: String, request: SendTransactionRequest): TransactionResponse =
        client.post("$baseUrl/wallets/$fromDid/send") {
            contentType(ContentType.Application.Json)
            setBody(request)
        }.body()

    suspend fun listChannels(): ChannelListResponse =
        client.get("$baseUrl/channels").body()

    suspend fun getMessages(channelId: String, limit: Int = 50): MessageListResponse =
        client.get("$baseUrl/channels/$channelId/messages?limit=$limit").body()

    suspend fun sendMessage(channelId: String, request: SendMessageRequest): MessageResponse =
        client.post("$baseUrl/channels/$channelId/messages") {
            contentType(ContentType.Application.Json)
            setBody(request)
        }.body()

    suspend fun createPost(request: CreatePostRequest): FeedEvent =
        client.post("$baseUrl/feed") {
            contentType(ContentType.Application.Json)
            setBody(request)
        }.body()
}

File: apps/ios/Nous/Sources/Network/NousAPI.swift
import Foundation

// MARK: - Response Types

struct HealthResponse: Codable {
    let status: String
    let version: String
    let uptimeMs: Int

    enum CodingKeys: String, CodingKey {
        case status, version
        case uptimeMs = "uptime_ms"
    }
}

struct IdentityResponse: Codable, Identifiable {
    var id: String { did }
    let did: String
    let displayName: String?
    let signingKeyType: String
    let exchangeKeyType: String

    enum CodingKeys: String, CodingKey {
        case did
        case displayName = "display_name"
        case signingKeyType = "signing_key_type"
        case exchangeKeyType = "exchange_key_type"
    }
}

struct BalanceEntry: Codable, Identifiable {
    var id: String { token }
    let token: String
    let amount: String
}

struct WalletResponse: Codable {
    let did: String
    let balances: [BalanceEntry]
    let nonce: Int
    let createdAt: String

    enum CodingKeys: String, CodingKey {
        case did, balances, nonce
        case createdAt = "created_at"
    }
}

struct TransactionResponse: Codable, Identifiable {
    let id: String
    let fromDid: String
    let toDid: String
    let token: String
    let amount: String
    let fee: String
    let memo: String?
    let status: String
    let timestamp: String

    enum CodingKeys: String, CodingKey {
        case id, token, amount, fee, memo, status, timestamp
        case fromDid = "from_did"
        case toDid = "to_did"
    }
}

struct TransferRequest: Codable {
    let fromDid: String
    let toDid: String
    let token: String
    let amount: Int
    let memo: String?

    enum CodingKeys: String, CodingKey {
        case token, amount, memo
        case fromDid = "from_did"
        case toDid = "to_did"
    }
}

struct ChannelResponse: Codable, Identifiable {
    let id: String
    let kind: String
    let name: String?
    let members: [String]
    let createdAt: String

    enum CodingKeys: String, CodingKey {
        case id, kind, name, members
        case createdAt = "created_at"
    }
}

struct MessageResponse: Codable, Identifiable {
    let id: String
    let channelId: String
    let sender: String
    let content: String
    let replyTo: String?
    let timestamp: String

    enum CodingKeys: String, CodingKey {
        case id, sender, content, timestamp
        case channelId = "channel_id"
        case replyTo = "reply_to"
    }
}

struct SendMessageRequest: Codable {
    let channelId: String
    let senderDid: String
    let content: String
    let replyTo: String?

    enum CodingKeys: String, CodingKey {
        case content
        case channelId = "channel_id"
        case senderDid = "sender_did"
        case replyTo = "reply_to"
    }
}

struct CreatePostRequest: Codable {
    let authorDid: String
    let content: String
    let replyTo: String?
    let hashtags: [String]?

    enum CodingKeys: String, CodingKey {
        case content, hashtags
        case authorDid = "author_did"
        case replyTo = "reply_to"
    }
}

struct FeedEvent: Codable, Identifiable {
    let id: String
    let pubkey: String
    let createdAt: String
    let kind: Int
    let content: String
    let tags: [[String]]

    enum CodingKeys: String, CodingKey {
        case id, pubkey, kind, content, tags
        case createdAt = "created_at"
    }
}

struct FeedResponse: Codable {
    let events: [FeedEvent]
    let count: Int
}

struct DaoResponse: Codable, Identifiable {
    let id: String
    let name: String
    let description: String
    let founderDid: String
    let memberCount: Int
    let createdAt: String

    enum CodingKeys: String, CodingKey {
        case id, name, description
        case founderDid = "founder_did"
        case memberCount = "member_count"
        case createdAt = "created_at"
    }
}

struct DaoListResponse: Codable {
    let daos: [DaoResponse]
    let count: Int
}

struct ProposalResponse: Codable, Identifiable {
    let id: String
    let daoId: String
    let title: String
    let description: String
    let proposerDid: String
    let status: String
    let createdAt: String
    let votingStarts: String
    let votingEnds: String
    let quorum: Double
    let threshold: Double

    enum CodingKeys: String, CodingKey {
        case id, title, description, status, quorum, threshold
        case daoId = "dao_id"
        case proposerDid = "proposer_did"
        case createdAt = "created_at"
        case votingStarts = "voting_starts"
        case votingEnds = "voting_ends"
    }
}

struct ProposalListResponse: Codable {
    let proposals: [ProposalResponse]
    let count: Int
}

// MARK: - API Client

final class NousAPI: @unchecked Sendable {
    static let shared = NousAPI()

    private let baseURL: String
    private let session: URLSession
    private let decoder: JSONDecoder

    init(baseURL: String = "http://localhost:8080/api/v1") {
        self.baseURL = baseURL
        self.session = URLSession.shared
        self.decoder = JSONDecoder()
    }

    private func get<T: Decodable>(_ path: String) async throws -> T {
        guard let url = URL(string: "\(baseURL)\(path)") else {
            throw URLError(.badURL)
        }
        let (data, _) = try await session.data(from: url)
        return try decoder.decode(T.self, from: data)
    }

    private func post<T: Decodable>(_ path: String, body: some Encodable) async throws -> T {
        guard let url = URL(string: "\(baseURL)\(path)") else {
            throw URLError(.badURL)
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)
        let (data, _) = try await session.data(for: request)
        return try decoder.decode(T.self, from: data)
    }

    // MARK: - Health

    func health() async throws -> HealthResponse {
        try await get("/health")
    }

    // MARK: - Identity

    func createIdentity(displayName: String? = nil) async throws -> IdentityResponse {
        struct Req: Encodable {
            let display_name: String?
        }
        return try await post("/identities", body: Req(display_name: displayName))
    }

    func getIdentity(did: String) async throws -> IdentityResponse {
        try await get("/identities/\(did)")
    }

    // MARK: - Wallet & Payments

    func getWallet(did: String) async throws -> WalletResponse {
        try await get("/wallets/\(did)")
    }

    func getTransactions(did: String, limit: Int = 50) async throws -> [TransactionResponse] {
        try await get("/wallets/\(did)/transactions?limit=\(limit)")
    }

    func createTransfer(request: TransferRequest) async throws -> TransactionResponse {
        try await post("/transfers", body: request)
    }

    // MARK: - Social Feed

    func getFeed(limit: Int = 20) async throws -> FeedResponse {
        try await get("/feed?limit=\(limit)")
    }

    func createPost(request: CreatePostRequest) async throws -> FeedEvent {
        try await post("/events", body: request)
    }

    // MARK: - Messaging

    func listChannels(did: String) async throws -> [ChannelResponse] {
        try await get("/channels?did=\(did)")
    }

    func getChannelMessages(channelId: String, limit: Int = 50) async throws -> [MessageResponse] {
        try await get("/channels/\(channelId)/messages?limit=\(limit)")
    }

    func sendMessage(request: SendMessageRequest) async throws -> MessageResponse {
        try await post("/messages", body: request)
    }

    // MARK: - Governance

    func listDaos() async throws -> DaoListResponse {
        try await get("/daos")
    }

    func listProposals() async throws -> ProposalListResponse {
        try await get("/proposals")
    }
}
